xdp-dump/xdpdump.c

```diff
@@ -0,0 +1,628 @@
+};
+
+bool    exit_xdpdump;
+pcap_t *exit_pcap;
+
+/*****************************************************************************
+ * snprinth()
+ *****************************************************************************/
+#define SNPRINTH_MIN_BUFFER_SIZE sizeof("0xffff:  00 11 22 33 44 55 66 77 88" \
+					" 99 aa bb cc dd ee ff	" \
+					"................0")
+
+static int snprinth(char *str, size_t size,
+		    const uint8_t *buffer, size_t buffer_size, size_t offset)
+{
+	int i;
+	int pre_skip;
+	int post_skip;
+	int zero_offset;
+
+	if (str == NULL || size < SNPRINTH_MIN_BUFFER_SIZE ||
+	    buffer == NULL || offset >= buffer_size || buffer_size > 0xffff)
+		return -EINVAL;
+
+	zero_offset = offset & ~0xf;
+	pre_skip = offset & 0xf;
+	post_skip = (zero_offset + 0xf) < buffer_size ? \
+		0 : 16 - (buffer_size - zero_offset);
+
+	/* Print offset */
+	snprintf(str, size, "0x%04zx:  ", offset & 0xfff0);
+	str += 9;
+
+	/* Print hex values */
+	if (pre_skip) {
+		memset(str, ' ', pre_skip * 3);
+		str[pre_skip * 3] = 0;
+	}
+
+	for (i = pre_skip; i < 16 - post_skip; i++) {
+		snprintf(str + (i * 3), 5, "%02x ",
+			 buffer[zero_offset + i]);
+	}
+
+	if (post_skip) {
+		memset(str + (i * 3), ' ', post_skip * 3);
+		str[(i * 3) + (post_skip * 3)] = 0;
+	}
+
+	/* Print printable chars */
+	str += 16 * 3;
+	*str++ = ' ';
+
+	if (pre_skip) {
+		memset(str, ' ', pre_skip);
+		str[pre_skip] = 0;
+	}
+	for (i = pre_skip; i < 16 - post_skip; i++)
+		str[i] = isprint(buffer[zero_offset + i]) ? \
+			buffer[zero_offset + i]: '.';
+
+	str[i] = 0;
+	return 0;
+}
+
+/*****************************************************************************
+ * handle_perf_event()
+ *****************************************************************************/
+static enum bpf_perf_event_ret handle_perf_event(void *private_data,
+						 int cpu,
+						 struct perf_event_header *event)
+{
+	uint64_t ts;
+	struct perf_handler_ctx *ctx = private_data;
+	struct perf_sample_event *e = container_of(event,
+						   struct perf_sample_event,
+						   header);
+	struct perf_lost_event *lost = container_of(event,
+						    struct perf_lost_event,
+						    header);
+
+	switch(e->header.type) {
+	case PERF_RECORD_SAMPLE:
+
+		if (e->header.size < sizeof(struct perf_sample_event) ||
+		    e->size < (sizeof(struct pkt_trace_metadata) + e->metadata.cap_len))
+			return LIBBPF_PERF_EVENT_CONT;
+
+		ts = e->time + ctx->epoch_delta;
+		if (ctx->pcap_dumper) {
+			struct pcap_pkthdr h;
+
+			h.ts.tv_sec = ts / 1000000000ULL;
+			h.ts.tv_usec = ts % 1000000000ULL / 1000;
+			h.caplen = min(e->metadata.cap_len, ctx->cfg->snaplen);
+			h.len = e->metadata.pkt_len;
+			pcap_dump((u_char *) ctx->pcap_dumper, &h,
+				  e->packet);
+		} else {
+			int i;
+			char hline[SNPRINTH_MIN_BUFFER_SIZE];
+			bool fexit = e->metadata.flags & MDF_DIRECTION_FEXIT;
+
+			printf("%llu.%09lld: packet size %u bytes, "
+			       "captured %u bytes on "
+			       "if_index %u, rx queue %u, @%s\n",
```

Comment:
The `@entry`/`@exit` was confusing to me the first time I ran this. Also, on exit we should also print the verdict; otherwise it will be confusing if people run this on a packet filter and they don't see any packets getting filtered. I also think we could make the output a bit shorter... 

So maybe change the output format to something like:

`1584701151.122081500: @entry: pkt size 98 bytes, captured 98 bytes on if_index 4, rx queue 5`
and
`1584701151.122081500: @exit[DROP]: pkt size 98 bytes, captured 98 bytes on if_index 4, rx queue 5`

(obviously we would like this in tcpdump output as well, but I guess we'll have to wait for pcapng support for that).

I'm also wondering if we should have a 'verbose mode' and only show the packet data dump if it is enabled. I.e., without verbose mode we would just do:
`1584701151.122081500: @exit[DROP]: pkt size 98 bytes on if_index 4, rx queue 5`

and with verbose mode we'd add the capture len and hex dump?